docs/use-cases.rst

```diff
@@ -0,0 +1,668 @@
+network, or a list of networks with known instances of a given |ContractType|.
+
+
+.. uml::
+
+  left to right direction
+  :Smart Contract Developer: as :Developer:
+
+  rectangle Querying {
+    (Query known instances on a given network) as (QueryNetwork)
+    (Query all networks for a given type) as (QueryType)
+    (Query all networks for instances with a given role) as (QueryInstance)
+    (Query for instance by network and address) as (QueryAddress)
+    (Query all contracts with a given bytecode) as (QueryBytecode)
+    (Query for bytecode ignoring link references) as (QueryBytecodeUnlinked)
+
+    QueryInstance .|> QueryType
+    QueryBytecodeUnlinked .|> QueryBytecode
+  }
+
+  Developer -- QueryNetwork
+  Developer -- QueryType
+  Developer -- QueryInstance
+  Developer -- QueryAddress
+  Developer -- QueryBytecode
+
+
+Query for instance by network and address
+`````````````````````````````````````````
+
+Users must be able to look up known |ContractInstances| given a |Network|
+and an |Address|.
+
+Query all contracts with a given bytecode
+`````````````````````````````````````````
+
+Because it is immutable, contract |Bytecode| serves as a reliable secondary
+index for both |ContractInstances| and |ContractTypes|. Users should be able to
+find all known |Contracts| based on the raw hexadecimal representation of
+EVM machine code.
+
+Query for bytecode ignoring link references
+```````````````````````````````````````````
+
+Two sets of contract |Bytecode| can differ only by |LinkValues|. Users should
+be able to query for all |ContractTypes| and |ContractInstances| matching
+a given bytecode, whether or not the |LinkReferences| have values or match.
+
+Query known instances on a given network
+````````````````````````````````````````
+
+Users should be able to see, at a glance, all |ContractInstances| on a given
+|Network|. This can be useful for validating migration state, or for easy
+listing of address / ABI information, to present/share externally.
+
+Query all networks for a given type
+```````````````````````````````````
+
+Developers often write applications so that each |ContractType| has a singleton
+deployed |ContractInstance|. Tooling should enable this first-class nature of
+contract types, and users should be able to track the various instances across
+all known |Networks|.
+
+This is particularly useful when creating front-end applications, referencing
+instances across networks with a single type description.
+
+Query all networks for instances with a given role
+``````````````````````````````````````````````````
+
+In cases where a particular |ContractType| is not deployed as a singleton,
+tooling should provide a mechanism by which users can deploy multiple
+|ContractInstances|, identified in some way distinct from type (i.e., *role*)
+
+As a result, users should be able to use this distinct identifier to query
+for analogous instances across |Networks|, instead of relying solely on the
+type.
+
```

Comment:
Is an analogous instance an instance of a contract type instantiated with specific constructor values? Or more subjective than that? Or are there other possible inputs like link addresses?

Reply:
Maybe this needs a more formal definition. In my head I've been imagining that analogous instances are instances with the same _role_, which would be a user-defined name, a là:
```javascript
deployer.deploy(HumanStandardToken, {as: "GoldToken"})
```.

Reply:
Is it worth another use case, to query instances with specified constructor args?

Reply:
@gnidan Oh I guess I was thinking about a network query - e.g. I have something I'm calling GoldToken and it specific traits apart from the name I gave it that would allow me to find analogous instances by trawling the network.

Reply:
@cgewecke yep, but I think constructor args are not sufficient to distinguish - it's possible that `GoldToken` and `SilverToken` have distinct roles but use the same constructor args.

Reply:
👍